substrate/primitives/staking/src/offence.rs

```diff
@@ -37,29 +37,6 @@ pub type Kind = [u8; 16];
 /// so that we can slash it accordingly.
 pub type OffenceCount = u32;
 
-/// In case of an offence, which conditions get an offending validator disabled.
-#[derive(
```

Comment:
I agree that a disablement strategy per offence doesn't make sense in the new case and we should remove it.

However, how about implementing a new offence strategy for the whole runtime? In this case, I could think of a type in the pallet config that implements a new trait `trait OffenseStrategy` or similar, which can implement the whole disablement strategy. The whole logic of which validators to disable, when, which to swap if the limit has been reached (perhaps the limits should be a CONST param in the trait too, etc).

The staking pallet or another type then could implement that trait in our runtime.

This way we could generalize this logic and maybe simplify the code.  wdyt?

Reply:
I assume your idea is any user of the pallet to be able to modify the disabling strategy to fit its needs.

I guess we can do this but:
* Disabling for a whole era can't be overridden ([here](https://github.com/paritytech/polkadot-sdk/blob/06e79d9723d08d29391ae53dcd0fcda2327d4c66/substrate/frame/staking/src/pallet/impls.rs#L429) and [here](https://github.com/paritytech/polkadot-sdk/blob/06e79d9723d08d29391ae53dcd0fcda2327d4c66/substrate/frame/staking/src/pallet/impls.rs#L512))
* Decision if an offender will be disabled or not can be part of the trait implementation you are suggesting ([here](https://github.com/paritytech/polkadot-sdk/blob/9f8c5986cc80ce4d11cb4a42b285a7b209481e82/substrate/frame/staking/src/slashing.rs#L339))
* `DisabledValidators` will remain part of the staking pallet. [This mutation](https://github.com/paritytech/polkadot-sdk/blob/9f8c5986cc80ce4d11cb4a42b285a7b209481e82/substrate/frame/staking/src/slashing.rs#L326) won't be changed besides the decision if a validator will be disabled or not (previous point).

Do I understand your idea correctly?

This might make the implementation a bit more modular. At this point I'm not sure if it will be simpler or not. But the practical benefits for the user will be not to disable a validator for some reason. Is this really useful considering that the disabling limit will be parametrized?

Reply:
> This might make the implementation a bit more modular. 

Yes, this is the idea! No changes to the logic itself, just making the disablement strategy abstracted and configurable. This may also be useful in the case on the staking parachain (we will re-use this pallet), which will probably have the disablement strategy split between the relay chain and the parachain (due to the delays of XCM messages from relay chain to parachain).

And also useful to other runtimes that use this pallet.

Reply:
The disabling strategy is implemented [via a trait](https://github.com/paritytech/polkadot-sdk/pull/2226/files#diff-e88e18be93824b5e9d452a4a409a7d7982d3ae53508c7737d27269c200d0da0eR1234) now. It has got a single function `make_disabling_decision` which returns a struct `DisablingDecision`.

I think this approach makes extending the disabling logic easier by extending the trait and the struct. I see no point to add more functionality which won't be used at the moment,